### ISSUE TITLE
Tracking issue for `cfg_select` (formerly `cfg_match`)

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust#115585]
Provides a native way to easily manage multiple conditional flags without having to rewrite each clause multiple times.

### Public API

```rust
cfg_select! {
    unix => {
        fn foo() { /* unix specific functionality */ }
    }
    target_pointer_width = "32" => {
        fn foo() { /* non-unix, 32-bit functionality */ }
    }
    _ => {
        fn foo() { /* fallback implementation */ }
    }
}

let is_unix_str = cfg_select! {
    unix => "unix",
    _ => "not unix",
};
println!("{is_unix_str}");
```

### Steps / History

- [x] Implementation (https://github.com/rust-lang/rust/pull/115416)
    - [x] Rename from `cfg_match` to `cfg_select` https://github.com/rust-lang/rust/pull/137198
    - [x] Support unbraced expressions: https://github.com/rust-lang/rust/pull/145233
- [ ] Final comment period
- [ ] Stabilization PR

### Unresolved Questions

* ~~What should the final syntax be? A match-like syntax feels more natural in the sense that each macro fragment resembles an arm.~~
* ~~Should the macro be supported directly by a language feature?~~
* ~~What should the feature name be? `cfg_match` conflicts with the already existing `cfg_match` crate.~~
* ~~How can we support usage in both expression-position and item position?~~
* ~~Support trailing commas to have similar grammar as `match` statements.~~ (#145233)

### References

* https://crates.io/crates/cfg-match
* https://crates.io/crates/cfg-if
* https://github.com/rust-lang/rust/issues/59442
* https://internals.rust-lang.org/t/is-cfg-if-going-to-be-pulled-into-std/13316

### Implementation history

- https://github.com/rust-lang/rust/pull/116312
- https://github.com/rust-lang/rust/pull/116323
- https://github.com/rust-lang/rust/pull/116342
- https://github.com/rust-lang/rust/pull/117162
- https://github.com/rust-lang/rust/pull/130313
- https://github.com/rust-lang/rust/pull/133720
- https://github.com/rust-lang/rust/pull/135625
- https://github.com/rust-lang/rust/pull/136261
- https://github.com/rust-lang/rust/pull/137198
- https://github.com/rust-lang/rust/pull/138993
- https://github.com/rust-lang/rust/pull/138996
- https://github.com/rust-lang/rust/pull/143941
- https://github.com/rust-lang/rust/pull/145233

#### Tooling support

- `rustfmt` side:
	- https://github.com/rust-lang/rustfmt/pull/6522
	- https://github.com/rust-lang/rust/pull/144323

**Comment by madsmtm** [comment]
I think it needs to be considered how this will interact with [`doc_auto_cfg`](https://github.com/rust-lang/rust/issues/43781).

Also, if this is intended to be `match`-like, then it seems weird to me that you can omit a fallback implementation? Unless of course all the possible combinations are statically known to be accounted for (which is hard, I know).

**Comment by CAD97** [comment]
Modulo (auto) doc(cfg), this implementation is more complex than it needs to be, since it's only supporting items.

<details><summary>simpler impl</summary>

```rust
#[macro_export]
macro_rules! cfg_match {
    {
        cfg($cfg:meta) => { $($cfg_items:item)* }
        $(cfg($rest_cfg:meta) => { $($rest_cfg_items:item)* })*
        $(_ => { $($fallthrough_items:item)* })?
    } => {
        #[cfg($cfg)]
        $crate::cfg_match! {
            _ => {
                $($cfg_items)*
                // validate the inactive cfgs as well
                #[cfg(any($($rest_cfg),*))]
                const _: () = ();
            }
        }
        #[cfg(not($cfg))]
        $crate::cfg_match! {
            $(cfg($rest_cfg) => { $($rest_cfg_items)* })*
            $(_ => { $($fallthrough_items)* })?
        }
    };
    {
        $(_ => { $($items:item)* })?
    } => {
        $($($items)*)?
    };
}
```

No internal matching arms necessary; just a normal inductive definition over the input syntax. The one tricky difference is that without the extra `#[cfg] const` item in the positive arm, later cfg predicates wouldn't be validated where they would be with the current implementation which always builds the full `#[cfg(all(yes, not(any(...))))]` matrix out.

This can be rather simply adapted to expression position by matching a block and using `break` with value.

-----

</details>

It's unfortunate though that implemented without compiler support `cfg_match!` is effectively limited to either supporting item position or expression position. Supporting both needn't block supporting only item position (the more necessary position, since expression context can use `if cfg!() else if` chains most of the time), but it'd be nice to ensure there's a path forward for a compiler enhanced version which can expand differently between item and expression position, so the macro can be made to just work on both.

**Comment by madsmtm** [comment]
Something else to consider: This should be able to detect invalid cfgs like the following (ideally regardless of whether the `cfg` is currently enabled or not):
```rust
match_cfg! {
    cfg(unix) => {}
    cfg(unix) => {} // Ooops, specified same `cfg` twice
    _ => {}
}
```

**Comment by thomcc** [comment]
I would prefer if it could support both items and expressions. It being part of the stdlib means that implementing it in the compiler would be completely reasonable, and would also make a lint that warns on redundant cfg arms pretty doable.

**Comment by jhpratt** [comment]
For detecting duplicates, it's important to note that this is more difficult than it seems at first. rustdoc has been dealing with this for `doc(cfg)`.

**Comment by CAD97** [comment]
I tested the current (1.74.0-nightly (e6aabe8b3 2023-09-26)) behavior of doc_auto_cfg: items defined in `match_cfg!` do not receive a portability info box. Theoretically cfg dependencies could be tracked through macro expansion, but that's not done currently; cfg dependencies are only recorded on items.

Fixing this is actually fairly simple once you understand why this is: instead of expanding to `#[cfg($cfg)] identity! { $($items)* }`, splat the cfg directly onto each item with `$(#[cfg($cfg)] $items)*`. The `match_cfg` crate does this; the `cfg_if` crate doesn't. A quick scan through the library directory didn't reveal any places where splatting cfg through `cfg_if!` would change documentation.

I want to additionally point out that doc_auto_cfg fundamentally can only see what items are defined on the current cfg and mark them as only available on whatever cfg *this specific implementation* is gated to. If you want to expose a portable API, either define a single portable public version of the API (and use private cfg gates to implement it) or use `doc(cfg())` to explicitly tell rustdoc what gate it should display as.

<details><summary>So what I personally think the macro definition wants to be:</summary>

<p>

<details><summary>in item position</summary>

```rust
#[macro_export]
macro_rules! cfg_match {
    // base case without wildcard arm
    () => {};

    // base case with wildcard arm (items)
    (_ => { $($wild_item:item)* }) => {
        $($wild_item)*
    };

    // process first arm (items; propagate cfg)
    (
        cfg($this_cfg:meta) => { $($this_item:item)* }
        $(cfg($rest_cfg:meta) => { $($rest_item:item)* })*
        $(_ => { $($wild_item:item)* })?
    ) => {
        $(#[cfg($this_cfg)] $this_item)*
        $crate::cfg_match! {
            $(cfg($rest_cfg) => { $(#[cfg(not($this_cfg))] $rest_item)* })*
            $(_ => { $(#[cfg(not($this_cfg))] $wild_item)* })?
        }
    };
}
```

</details>

<details><summary>in expression position</summary>

```rust
#[macro_export]
macro_rules! cfg_match {
    // base case without wildcard arm
    () => {};

    // base case with wildcard arm (expression block)
    (_ => $wild:block) => {
        $wild
    };

    // process first arm (expression block)
    (
        cfg($this_cfg:meta) => $this:block
        $(cfg($rest_cfg:meta) => $rest:block)*
        $(_ => $wild:block)?
    ) => {
        {
            #[cfg($this_cfg)]
            $crate::cfg_match! {
                _ => $this
            }
            #[cfg(not($this_cfg))]
            $crate::cfg_match! {
                $(cfg($rest_cfg) => $rest)*
                $(_ => $wild)?
            }
        }
    };
}
```

</details>

<details><summary>in statement position</summary>

```rust
#[macro_export]
macro_rules! cfg_match {
    // base case without wildcard arm
    () => {};

    // base case with wildcard arm
    (_ => { $($wild:tt)* }) => {
        $($wild)*
    };

    // process first arm
    (
        cfg($this_cfg:meta) => { $($this:tt)* }
        $(cfg($rest_cfg:meta) => { $($rest:tt)* })*
        $(_ => { $($wild:tt)* })?
    ) => {
        #[cfg($this_cfg)]
        $crate::cfg_match! {
            _ => { $($this)* }
        }
        #[cfg(not($this_cfg))]
        $crate::cfg_match! {
            $(cfg($rest_cfg) => { $($rest)* })*
            $(_ => { $($wild)* })?
        }
    };
}
```

This technically is the most flexible spelling just passing through tt sequences; it works perfectly for items and statements (modulo preserving doc_auto_cfg) and can even be used for expressions if wrapped inside an outer block (e.g. `{ cfg_match! { ... } }`).

</details>

-----

</details>

There's also the further question of how much syntax checking is expected for inactive arms. The current implementation that matches everything as `$($:item)*` does full syntax validation, but a naive implementation forwarding tokens context-antagonistically as `$($:tt)*` can easily cfg out tokens without ever placing them in a syntax-validated position -- in fact the above statement macro form does this.

<details><summary>A far too overly clever implementation:</summary>

```rust
#[macro_export]
macro_rules! cfg_match {
    // base case (wildcard arm)
    () => {};

    // base case (no wildcard arm)
    (_ => { $($wild:tt)* }) => {
        $($wild)*
    };

    // Note: macro patterns matching $:stmt need to come before those matching
    // $:item, as otherwise statements will match as the $:item pattern instead
    // even when not valid as items (and obviously thus cause a syntax error).

    // apply cfg to wildcard arm (expression position)
    (_ => #[cfg($cfg:meta)] { $wild:expr }) => {
        { #[cfg($cfg)] { $wild } }
    };

    // apply cfg to wildcard arm (statement position)
    (_ => #[cfg($cfg:meta)] { $wild_stmt:stmt; $($wild:tt)* }) => {
        #[cfg($cfg)]
        $crate::cfg_match! {
            _ => { $wild_stmt; $($wild)* }
        }

        // We only parse the first statement in the macro pattern, so we need to
        // emit the captured syntax even when the cfg is inactive such that any
        // syntax errors still get reported. We do the macro capture this way as
        // if we match as statements, minor typos try to parse as items instead.
        #[cfg(any())]
        const _: () = { wild_stmt; $($wild)* };
    };

    // apply cfg to wildcard arm (item position)
    (_ => #[cfg($cfg:meta)] { $($wild:item)* }) => {
        $(#[cfg($cfg)] $wild)*
    };

    // merge multiple cfg into a single cfg(all()) predicate
    (_ => $(#[cfg($cfg:meta)])+ { $($wild:tt)* }) => {
        $crate::cfg_match! {
            _ => #[cfg(all($($cfg),+))] { $($wild)* }
        }
    };

    // split off first arm (empty)
    (
        cfg($this_cfg_arm:meta) => $(#[cfg($this_cfg_pre:meta)])* { /* empty */ }
        $(cfg($rest_cfg_arm:meta) => $(#[cfg($rest_cfg_pre:meta)])* { $($rest:tt)* })*
        $(_ => $(#[cfg($wild_cfg_pre:meta)])* { $($wild:tt)* })?
    ) => {
        $crate::cfg_match! {
            $(cfg($rest_cfg_arm) => $(#[cfg($rest_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $($rest)* })*
            $(_ => $(#[cfg($wild_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $($wild)* })?
        }
    };

    // split off first arm (expression position)
    (
        cfg($this_cfg_arm:meta) => $(#[cfg($this_cfg_pre:meta)])* { $this:expr }
        $(cfg($rest_cfg_arm:meta) => $(#[cfg($rest_cfg_pre:meta)])* { $rest:expr })*
        $(_ => $(#[cfg($wild_cfg_pre:meta)])* { $wild:expr })?
    ) => {{
        $crate::cfg_match! {
            _ => $(#[cfg($this_cfg_pre)])* #[cfg($this_cfg_arm)] { $this }
        }
        $crate::cfg_match! {
            $(cfg($rest_cfg_arm) => $(#[cfg($rest_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $rest })*
            $(_ => $(#[cfg($wild_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $wild })?
        }
    }};

    // split off first arm (statement position)
    (
        cfg($this_cfg_arm:meta) => $(#[cfg($this_cfg_pre:meta)])* { $this_stmt:stmt; $($this:tt)* }
        $(cfg($rest_cfg_arm:meta) => $(#[cfg($rest_cfg_pre:meta)])* { $($rest:tt)* })*
        $(_ => $(#[cfg($wild_cfg_pre:meta)])* { $($wild:tt)* })?
    ) => {
        $crate::cfg_match! {
            _ => $(#[cfg($this_cfg_pre)])* #[cfg($this_cfg_arm)] { $this_stmt; $($this)* }
        }
        // Ensure all arms infer statement position by prefixing a statement. We
        // only match the first arm in the macro pattern because otherwise minor
        // typos unhelpfully cause all arms to parse as item definitions instead
        // and thus reporting an error nowhere near the actual problem, or even
        // more amusingly, accidentally providing automatic semicolon insertion.
        $crate::cfg_match! {
            $(cfg($rest_cfg_arm) => $(#[cfg($rest_cfg_pre)])* #[cfg(not($this_cfg_arm))] { {}; $($rest)* })*
            $(_ => $(#[cfg($wild_cfg_pre)])* #[cfg(not($this_cfg_arm))] { {}; $($wild)* })?
        }
    };

    // split off first arm (item position)
    (
        cfg($this_cfg_arm:meta) => $(#[cfg($this_cfg_pre:meta)])* { $($this:item)* }
        $(cfg($rest_cfg_arm:meta) => $(#[cfg($rest_cfg_pre:meta)])* { $($rest:item)* })*
        $(_ => $(#[cfg($wild_cfg_pre:meta)])* { $($wild:item)* })?
    ) => {
        $crate::cfg_match! {
            _ => $(#[cfg($this_cfg_pre)])* #[cfg($this_cfg_arm)] { $($this)* }
        }
        // Items just match as item in the macro pattern. The error messages
        // would be improved if we could pass through tokens without eagerly
        // matching them as items, but we need to parse the items in the macro
        // so that we can apply the cfg attributes to every item in the arm.
        $crate::cfg_match! {
            $(cfg($rest_cfg_arm) => $(#[cfg($rest_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $($rest)* })*
            $(_ => $(#[cfg($wild_cfg_pre)])* #[cfg(not($this_cfg_arm))] { $($wild)* })?
        }
    };
}
```

This does _almost_ always behave as desired by using the arm bodies to try to guess how to expand, but is unreasonably complicated by attempting to do so and as a result causes some very poor errors for reasonable syntax errors.

</details>

**Comment by c410-f3r** [comment]
Well, since https://github.com/rust-lang/rust/pull/115416#issuecomment-1707457225 I have been working in transposing `cfg_match` to a built-in in order to easily allow single elements in an arm without brackets. Such thing will also permit any other feature that is being discussed here AFAICT.

It is worth noting that the current implementation mimics `cfg_if` and allows the expansion of token trees because `cfg`s are applied to macro calls as @CAD97 previously commented.

```rust
cfg_match! {
    cfg(unix) => { fn foo() -> i32 { 1 } },
    _ => { fn foo() -> i32 { 2 } },
}

# Expands to ->

#[cfg(all(unix, not(any())))]
cfg_match! { @__identity fn foo() -> i32 { 1 } }

#[cfg(all(not(any(unix))))]
cfg_match! { @__identity fn foo() -> i32 { 2 } }
```

That said, changing the strategy to apply `cfg` on each element inside a block **will** trigger `stmt_expr_attributes` (https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4cf4facc19b052053cca655b2b42f782) for expressions and conceivably/possibly/maybe/perhaps, a large set of elements along a large set of `cfg`s could affect compilation times. **But**, it seems to be the correct approach due to `doc_auto_cfg`.

**Comment by c410-f3r** [comment]
The syntax of `select`-like macros differ from the syntax of `match` statements so we have an inconsistency but personally, I don't really mind moving forward with "cfg_select" or "select_cfg".

At least for me, it became clear in https://github.com/rust-lang/rust/pull/116342 that `cfg` prefixes feel unnecessary, verbose and redundant. The following syntax would be much better in my opinion.

```rust
match_cfg! {
    unix => {
        fn foo() {}
    }
    target_pointer_width = "32" => {
        fn foo() {}
    }
    _ => {
        fn foo() {}
    }
}
```

The functionality provided by this feature is useful and can positively impact several projects. If possible, it would be awesome to seek consensus as soon as possible to create a path for stabilization.

**Comment by the8472** [comment]
> Unresolved Questions
>    How can we support usage in both expression-position and item position?

At least in std we rely on both, cfg-if'd sections in function bodies and defining items, e.g. platform-specific modules. See  https://github.com/rust-lang/rust/pull/126827#discussion_r1650198355 for an example where platform-dependent definitions + other macros + ordering constraints on  interacted poorly enough that outsourcing things to a module would have been tricky.

**Comment by joshtriplett** [comment]
I tested this, and trailing commas seem to work:

```rust
cfg_match! {
    unix => {
        fn f1() {
            println!("unix");
        }
    },
    _ => {
        fn f1() {
            println!("not unix");
        }
    },
}
```

It also works in function bodies, for whole statements.

```rust
fn f2() {
    cfg_match! {
        unix => {
            println!("unix");
        },
        _ => {
            println!("not unix");
        },
    }
}
```

It doesn't seem to work in expression position, but that's something that could be added compatibly in the future.

Based on all of that, I *think* this is ready to consider stabilizing:

@rfcbot merge

**Comment by Skgland** [comment]
> It doesn't seem to work in expression position, but that's something that could be added compatibly in the future.
> 

It appears to work as a blocks tail expression though, but not as a general expression: [play-ground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2aee205faa4d03dc433c488e533ddc77)

**Comment by madsmtm** [comment]
It'd be nice if the tracking issue could be updated with places where the decisions for the unresolved questions were made?

---

I still feel like it's a mistake to stabilize this without compiler support, there's just so many things that are still undesirable about the current implementation:
- `doc_auto_cfg` doesn't work (because the `#[cfg]` attributes are not placed on items, but on the macro).
- The macro [does now](https://github.com/rust-lang/rust/pull/135625) support expression position, but it requires extra braces, which is quite confusing.
- The braces around the resulting item/expression are unnecessary. Ideally, we'd be able to do:
    ```rust
    let res = cfg_match! {
        unix => libc::foo(),
        _ => libc::bar(),
    };
    ```

I recognize that none of these are blocking concerns, but at the very least we should consider not stabilizing the expression position support, as that is certainly something a future compiler implementation would want to avoid.

**Comment by tgross35** [comment]
> I just found out about this today, and my first impression is that semantically this isn't really a "`match`"
>
> [...]
> 
> All that is to say that `cfg_select!` seems like a better name to me.

As-is, this seems worth changing; I would expect a macro named after `match` to support a lot of `match` syntax, but we don't. I opened https://github.com/rust-lang/rust/pull/137198 to do this.

Alternatively, maybe we should make this look and work like `match`. This could look like:

```rust
cfg_match! {
    unix => 0,           // Simple match
    unix | windows => 0, // One of many
    target_pointer_width: "32" => 0,        // Match a single `=` config value
    target_pointer_width: "16" | "32" => 0, // One of many `=` values
    feature: ["std", "derive"] => 0,        // `feature = "std"`, `feature = "derive"`, no other `feature`
    feature: ["std", "derive", ..] => 0,    // std, derive, and possibly others
    _ => 0
}
```

Which can effectively expand to something like this:

```rust
pub struct Cfg {
    unix: bool,
    windows: bool,
    target_pointer_width: &'static [&'static str],
    feature: &'static [&'static str]
}

match config {
    Cfg { unix: true, .. } => 0,
    Cfg { unix: true, .. } | Cfg { windows: true, .. } => 0,
    Cfg { target_pointer_width: ["16" | "32"], .. } => 0,
    // Ordering does not matter
    Cfg { feature: ["std", "derive"], .. } => 0,
    Cfg { feature: ["std", "derive", ..], .. } => 0,
    _ => 0
}
```

**Comment by tgross35** [comment]
What is the motivation for this being based on `match` rather than `if` in the first place? `match` has a much more complicated syntax and set of rules to replicate as a macro, this macro isn't really consistent with those rules now (I don't think it can be), and it seems unclear if we gain anything substantial in exchange for the complexity. The top post's example could just as easily be:

```rust
static_cfg! {
    if cfg(unix) {
        fn foo() {}
    } else if cfg(target_pointer_width = "32") {
        fn foo() {}
    } else {
        fn foo() {}
    }
}
```

(I am including the `cfg` wrapper because a visual separator between expression syntax and `cfg` predicates seems useful. E.g. `cfg(a && b)` is not valid, `cfg(a) && cfg(b)` could one day be added).

**Comment by c410-f3r** [comment]
Unbelievable, once again a feature has blocking concerns at the final stages of stabilization. It may be worth adding an official blog post for each new feature as well as each new substantial change to increase awareness.

For any following discussion, try to focus on a simple but extensible MVP for shipping purposes. Further improvements that require further elaborations could then be postponed to future versions.

Perfection is the enemy of good and everyone here knows the usefulness of this macro.

**Comment by jhpratt** [comment]
> Unbelievable, once again a feature has blocking concerns at the final stages of stabilization.

@c410-f3r That's literally the purpose of FCP.

**Comment by joshtriplett** [comment]
>     * `doc_auto_cfg` doesn't work (because the `#[cfg]` attributes are not placed on items, but on the macro).

I definitely wouldn't consider this a blocker, though it'd be nice if someone wanted to provide a fix.

>     * The macro [does now](https://github.com/rust-lang/rust/pull/135625) support expression position, but it requires extra braces, which is quite confusing.

>     * The braces around the resulting item/expression are unnecessary. Ideally, we'd be able to do:
>       let res = cfg_match! {
>           unix => libc::foo(),
>           _ => libc::bar(),
>       };

I agree, but that's something we could fix in the future.

> I recognize that none of these are blocking concerns, but at the very least we should consider not stabilizing the expression position support, as that is certainly something a future compiler implementation would want to avoid.

Can you clarify whether there's a backwards compatibility concern here, such that the compiler implementation would want to *remove* a feature relative to the current macro version?

If there's a backwards compatibility concern, happy to block this waiting on a solution. If there's no backwards compatibility concern, then I continue to think we should ship this and let it add more features in the future.

**Comment by hanna-kruppe** [comment]
> since this largely exists to (substantially) reduce typing, it seemed preferable to _further_ reduce typing by not requiring the user to repeatedly write `cfg()`.

I understand where this is coming from, but to me there is a big difference between these two kinds of "reducing typing". Reducing error-prone repetition of complex `cfg(..)` expressions because there's no `else` for `#[cfg]` is extremely useful because it can prevent subtle bugs, enough for me to happily depend on `cfg_if` in most non-trivial cases. In contrast, the benefit of not having to type the `cfg(` and `)` decorations is minimal to me, it might even mess up my muscle memory since they're needed in every other context. It also introduces an inconsistency with other places where the `cfg(..)` is spelled out even though context makes it redundant, like check-cfg ([in `rustc --check-cfg`](https://doc.rust-lang.org/rustc/check-cfg.html), [in the Cargo.toml lints table](https://doc.rust-lang.org/rustc/check-cfg/cargo-specifics.html#check-cfg-in-lintsrust-table), and [in build script output](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-check-cfg)) and the proposed https://github.com/rust-lang/rfcs/pull/3759.

**Comment by traviscross** [comment]
Those other cases you mention are interesting.  However, this is the first place where we'd be adding something like this to something that feels like part of the language itself, and that is a key distinction.  I'm not seeing a lot of value in redundancy here, and conversely, if we required the arm matchers to be explicitly `cfg(..)` wrapped, then it'd seem most consistent for the else case to be `cfg(_)` or `cfg(all())` or something, and that seems unpleasant.

Put another way, that `cfg_match!` arm matchers are in `cfg`-space feels just as implied as that `match` arm matchers are in pattern/type space (i.e. rather than expression-space).

cc @rust-lang/lang

**Comment by hanna-kruppe** [comment]
I don't have a strong preference personally, just wanted to note that I consider the motivation for this style of "reducing typing" different and smaller than the one motivating the feature as a whole.

Also, question for clarification since I didn't see it in the issue description or currently nightly docs: `cfg` "operators" like `all` and `not` would work the same as elsewhere, just with the outer `cfg(_)` removed? As in:

```rust
cfg_match! {
    all(target_arch = "x86_64", target_feature = "sse2") => { ... }
    all(target_arch = "aarch64", target_feature = "neon", not(target_endian = "big")) => { ... }
    not(any(target_has_atomic = "32", target_has_atomic = "64")) => { ... }
    _ => { ... }
}
```

**Comment by ChrisDenton** [comment]
Yes, see for example https://github.com/rust-lang/rust/blob/master/compiler%2Frustc_span%2Fsrc%2Fanalyze_source_file.rs

It's much closer to `#[cfg]` than a real `match`.

**Comment by CAD97** [comment]
I just want to note again that a "pattern" syntax of `key = "value" =>` syntax "feels" a bit more foreign than `cfg(key = "value")`, which "feels" more like a struct pattern in shape, even though it isn't. Because it very much isn't, the different feeling could be an advantage just as much as it could be a disadvantage.

As long as rustfmt has the ability to format the macro usage, though, I have no real preference between the two, beyond the familiarity angle. (Even if rustfmt fails to format in many/most cases; just adding the `match_cfg` grammar as one of the options tried for all parenthesized macros would be sufficient for me.)

----------

There are currently still a lot of uses of `cfg_if!` in std. If we're moving to stabilize `match_cfg!` and recommend that people use it, it probably makes sense to migrate to using it in std first.

**Comment by jplatte** [comment]
Not sure whether there are any forward-compat issues with adding this later, but when I tried to use this (the crates.io crate), I pretty quickly ran into not being able to conditionally introduce `let` bindings into the surrounding scope, which `cfg_if!` supports. The [upstream issue](https://github.com/arcnmx/cfg-match-rs/issues/1) has not received any attention.

**Comment by ChrisDenton** [comment]
> There are currently still a lot of uses of `cfg_if!` in std. If we're moving to stabilize `match_cfg!` and recommend that people use it, it probably makes sense to migrate to using it in std first.

The trouble with std is that at the moment any changes mean we need to use `#[cfg(bootstrap)]` to include both the old and the new syntax because std is built with both the bootstrap compiler and with the current HEAD compiler. That will hopefully change in the future but it hasn't yet. Until that changes I think it would be good to be at least *reasonably* sure of the syntax we're stabilising before going all in on it. The compiler is generally a better place to test out features that are likely to change because they can be changed all at once without any fuss.

**Comment by CAD97** [comment]
`cfg(bootstrap)` is only necessary if changes are made to the compiler or if `realstd::` is used. `cfg_match!` is just another macro that's defined in `core`; using it would be no different from using `cfg_if!`, which is also defined to be used from within `core` as a copy of the crate provided macro. The only difference would be calling `cfg_if::cfg_if!` with its syntax or calling `crate::match_cfg!` with its syntax, and that rustfmt has special handling for `cfg_if!` but not `match_cfg!`.

**Comment by joshtriplett** [comment]
@jplatte wrote:
> Not sure whether there are any forward-compat issues with adding this later, but when I tried to use this (the crates.io crate), I pretty quickly ran into not being able to conditionally introduce `let` bindings into the surrounding scope, which `cfg_if!` supports. The [upstream issue](https://github.com/arcnmx/cfg-match-rs/issues/1) has not received any attention.

I believe that would indeed be a forward-compatibility concern. It'd be quite easy to construct code that would change meaning if that were introduced, by shadowing.

@rfcbot concern let-bindings

**Comment by CAD97** [comment]
The macro as currently formulated in core doesn't have any issues with introducing let bindings into the containing scope, actually. It's just the older version of the macro as provided by the linked crate that has that issue.

**Comment by tmandry** [comment]
This is very adjacent to a core language feature, and procedurally I think extending that feature with a macro in core should involve lang input.

@rustbot label I-lang-nominated

Personally I am very much in favor of improving the usability of `cfg`. We should consider whether the use cases of this macro are better addressed with language features, as in the discussions happening in https://github.com/rust-lang/rfcs/pull/3796 and https://github.com/rust-lang/rfcs/pull/3804, or how those RFCs would interact with this macro if both were accepted.

If we do accept this macro and it accepts novel syntax for matching against cfgs (like using `|` to match against different values), we should consider whether that is something we ought to accept in the language too.

**Comment by traviscross** [comment]
In that case...  I forget, does it let me...

@rfcbot cancel

**Comment by joshtriplett** [comment]
@tmandry To the best of my understanding, `cfg_match!` as *currently* implemented does not add any novel cfg syntax. I would expect that if and when we add extensions to the `cfg` syntax, we'll add the same extensions to `cfg_match!`, but I don't anticipate us adding `cfg_match!` extensions that aren't in `cfg`.

**Comment by joshtriplett** [comment]
@scottmcm and others in the lang meeting made the point that while we can't necessarily *enforce* exhaustiveness, we *might* want to consider the possibility of having an implicit `_ => compile_error!(...)` or `_ => compile_warning!(...)`, such that you have to write `_ => {}` if you really want it to expand to nothing in that case.

I don't know if we want to do that by *default*, or if we want that to be a *lint* of some kind, but either way, it's worth considering. @rust-lang/libs-api

**Comment by traviscross** [comment]
Based on having discussed this, and based on other recent related discussions such as the ones @tmandry mentions, I'm feeling more strongly than I did in the past that if we want to include `match` in the name, it should do what it can to simulate exhaustiveness, and so I'd prefer to see it give a compiler error if it reaches a case that is not covered by an arm.  That will, in practice I think, give the code that people write with this more of the "flavor" of a `match`, increasing language cohesiveness.

**Comment by traviscross** [comment]
@rustbot labels +I-libs-api-nominated

Nominating for libs-api to consider feedback from lang discussion.

**Comment by tmandry** [comment]
> To the best of my understanding, `cfg_match!` as _currently_ implemented does not add any novel cfg syntax. I would expect that if and when we add extensions to the `cfg` syntax, we'll add the same extensions to `cfg_match!`, but I don't anticipate us adding `cfg_match!` extensions that aren't in `cfg`.

Sounds good then.



> we _might_ want to consider the possibility of having an implicit `_ => compile_error!(...)` or `_ => compile_warning!(...)`, such that you have to write `_ => {}` if you really want it to expand to nothing in that case.

> if we want to include `match` in the name, it should do what it can to simulate exhaustiveness, and so I'd prefer to see it give a compiler error if it reaches a case that is not covered by an arm. That will, in practice I think, give the code that people write with this more of the "flavor" of a `match`, increasing language cohesiveness.

I'm somewhat persuaded by these. If there were a strong reason not to do this it would be okay not to, though I might also quibble with the name in that case. But it feels like a nice way to tie things together, while guarding against unplanned config combinations with an explicit error message.

**Comment by folkertdev** [comment]
I don't see this mentioned yet: I expect this to work (e.g. `concat!` does work in that position)

```rust
#![feature(cfg_match)]

fn main() {
    println!(cfg_match! {{
        unix => { "foo" }
        _ => { "bar" }
    }});
}
```

I believe given the current grammar https://doc.rust-lang.org/src/core/macros/mod.rs.html#330, the `{ ... }` on the right-hand side of the `=>` is required. 

But I believe that, in expression contexts at least, that expands to a `Expr::Block`. In the case of some builtin macros like `format!`, that hits this error branch:

https://github.com/rust-lang/rust/blob/5fe04cbebd593c6530a21cd7bd0e8e9503b1ffe4/compiler/rustc_builtin_macros/src/util.rs#L129

The quick fix is to accept a block containing a string literal, I guess? I don't think there is a better solution when using `macro_rules!` but the extra `Expr::Block` wrapper is inelegant.

**Comment by alice-i-cecile** [comment]
FYI, Bevy has added its own version of this in https://github.com/bevyengine/bevy/pull/18822, primarily to rein in virality of `std` feature flags. The PR description and comments are worth a read there, and I'd be happy to connect you with our contributors if you want user feedback on designs.

**Comment by traviscross** [comment]
On the libs-api call today, this was discussed at some length.  Here's where it landed:

- We want to name this `cfg_select!`.
- If an uncovered case is actually encountered, a compiler error is emitted.
- We don't want to stabilize the `cfg_select! {{ .. }}` (double braces) expression-position syntax.
- We don't want to stabilize it yet in expression position at all, unless it's wrapped in outer braces, e.g. `{ cfg_select! { .. } }`.

On the name, the more we talked about it, the more we continued to develop cold feet about calling this a kind of `match`.  The reasons are essentially those that have been discussed in this thread; they just kept adding up.  At the end of the day, it's not doing "matching", as Rust means it, it's evaluating a list of conditions, in order, and picking the first arm where the condition evaluates as true.

We committed to and engaged in a thorough syntactic bikeshed to pick a better name.  The two candidates that came out of that as plausible -- with general support and few objections -- were `cfg_cond` and `cfg_select`.

The downside of `cond` is that, while it suggests "conditional" and is used in other languages with the semantic here, it would be new and novel jargon for many Rust programmers.  The upside is that, by being new, it would have no baggage in terms of alternate common meanings in our ecosystem.

The upside of `select` is that it is more familiar, due to existing use in our ecosystem, and it reads as a verb.  The downside is that it does carry the baggage of alternate meanings from use in the ecosystem, such as Tokio's `select!`, where the macro has a different syntax and polls the future on each arm.

In the end, we leaned in favor of the name with more existing familiarity and that could be read as a verb, and went with `cfg_select`.

Not including `match` in the name removes the objections of a lang nature to this not being exhaustive.  However, based on general Rust principles of trying to push errors to the earliest possible point, many of us still felt that giving an error if an uncovered case is encountered remains a virtue.  Often, an uncovered case would end up simply leading to an error or undesired behavior later.  So on that basis, we'd like the `cfg_select!` macro to emit code that produces a compiler error upon reaching an uncovered case.

Currently, the macro supports an expression-position syntax that requires the use of double braces such as `cfg_select! {{ .. }}`.  People thought this was weird and nobody wanted to stabilize this, so this needs to be removed.

More generally, we felt that there were a number of open questions we should address regarding the behavior of this macro in expression position.

One, while we were OK with always requiring the arm body to be braced in item position, we were still discussing the merits of allowing unbraced arm bodies in expression position.

Two, currently, the macro always returns a `block`, and this creates problems for use in places such as operand position for inline assembly, as discussed in https://github.com/rust-lang/rust/issues/140279 and https://github.com/rust-lang/rust/pull/140367#discussion_r2072362396.  The way the macro returns the body is observable -- e.g., in Rust 2024, blocks have different behavior with respect to temporary lifetimes -- and so we should make sure the behavior is what we want prior to stabilizing in a way this could be observed.  Probably this macro will need to become a built-in compiler macro.

Some were fine with simply blocking on that implementation work being done.  Another option, though, is that we discussed that we'd be OK with stabilizing this only in item position and in expression position when wrapped in outer braces such as `{ cfg_select! { .. } }`.

Thanks to everyone in this thread for the many useful bits of analysis and feedback that were helpful in arriving at this point after some back and forth.

The plan is to cancel the current proposed FCP and to then restart it with the proposal as detailed in this comment.

**Comment by Amanieu** [comment]
@rfcbot cancel

**Comment by folkertdev** [comment]
Is this blocked on someone actually doing the work of making `cfg_select!` a builtin macro? I've been doing some work on the `asm!` macros recently so I'd be interested in moving the implementation forward.

**Comment by traviscross** [comment]
Yes, that's roughly correct (we [discussed](https://github.com/rust-lang/rust/issues/115585#issuecomment-2855291355) options for a partial stabilization short of doing that, but just doing this correctly would probably be better if possible), and if you could do that, it'd be much appreciated.

When touching this, I'd suggest also fixing the behavior (as compared to the current macro) so that it accepts invocations that only contain the uncovered case, e.g. `cfg_select! { _ => { todo!() } }`.  We had discussed that in earlier calls, and while nobody  wanted to block a partial stabilization on fixing that, since it  could be fixed later, that is something people thought should be accepted.

**Comment by Jules-Bertholet** [comment]
Is `_` fallback necessary? `true` does the same job, no?

**Comment by joshtriplett** [comment]
> Is `_` fallback necessary? `true` does the same job, no? And, as has been discussed previously, this isn’t a `match`.

`_` makes it clearer that this is the nothing-above-matched case.

**Comment by tgross35** [comment]
I think https://github.com/rust-lang/rust/issues/149090 may be nice to have, but it is not a blocker (`cfg_if` has the same problem).

**Comment by folkertdev** [comment]
I'd be in favor of FCP: there is a large upside for the ecosystem and I don't see the rustfmt stuff moving unfortunately.

**Comment by blyxyas** [comment]
Should `cfg_select` keep their expansion markers? We're having problems at Clippy because `cfg_select` results are indistinguishable of real code (and the `SyntaxContext` is identical), producing weird warnings.

See https://github.com/rust-lang/rust-clippy/issues/17498
